core/src/main/java/com/linecorp/armeria/server/cors/CorsPolicy.java

```diff
@@ -0,0 +1,300 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.cors;
+
+import static com.linecorp.armeria.server.cors.CorsService.ANY_ORIGIN;
+
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.common.DefaultHttpHeaders;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.server.cors.CorsConfig.ConstantValueSupplier;
+import com.linecorp.armeria.server.cors.CorsConfig.InstantValueSupplier;
+
+import io.netty.util.AsciiString;
+
+/**
+ * Contains information of the CORS policy with the specified origins.
+ */
+public final class CorsPolicy {
+
+    private static final Logger logger = LoggerFactory.getLogger(CorsPolicy.class);
+    private static final String DELIMITER = ",";
+    private static final Joiner HEADER_JOINER = Joiner.on(DELIMITER);
+    private final Set<String> origins;
+    private final boolean credentialsAllowed;
+    private final boolean nullOriginAllowed;
+    private final long maxAge;
+    private final Set<AsciiString> exposedHeaders;
+    private final Set<HttpMethod> allowedRequestMethods;
+    private final Set<AsciiString> allowedRequestHeaders;
+    private final String joinedExposedHeaders;
+    private final String joinedAllowedRequestHeaders;
+    private final String joinedAllowedRequestMethods;
+    private final HttpHeaders preflightResponseHeaders;
+
+    CorsPolicy(final Set<String> origins, boolean credentialsAllowed, long maxAge,
+               boolean nullOriginAllowed, Set<AsciiString> exposedHeaders,
+               Set<AsciiString> allowedRequestHeaders, Set<HttpMethod> allowedRequestMethods,
+               boolean preflightResponseHeadersDisabled,
+               Map<AsciiString, Supplier<?>> preflightResponseHeaders) {
+        this.origins = origins;
+        this.credentialsAllowed = credentialsAllowed;
+        this.maxAge = maxAge;
+        this.nullOriginAllowed = nullOriginAllowed;
+        this.exposedHeaders = ImmutableSet.copyOf(exposedHeaders);
+        this.allowedRequestMethods = EnumSet.copyOf(allowedRequestMethods);
```

Comment:
How about changing the type of the parameter `allowedRequestMethods` to `EnumSet<HttpMethod>` and using `ImmutableSet.copyOf()` here, so that Guava internally uses `ImmutableEnumSet`?